lab/src/raft/leader_election.go

```diff
@@ -0,0 +1,34 @@
+package raft
+
+import (
+	"time"
+)
+
+func (rf *Raft) toCandidate() {
+	rf.state = CANDIDATE
+	rf.currentTerm++  // Increment currentTerm
+	rf.votedFor = rf.me  // Vote for self
+}
+
+func (rf *Raft) runLeaderElection() {
+	rf.toCandidate()
+
+	replies := make([]RequestVoteReply, len(rf.peers))
+	voteChan := make(chan int, len(rf.peers) - 1)
+
+	// TODO(ling): Implement vote
+}
+
+// Kick off leader election periodically by sending out RequestVote RPCs
+// when it hasn't heard from another peer for a while.
+func (rf *Raft) startLeaderElection() {
+	rf.Lock()
+	defer rf.Unlock()
+
+	electionTimeout := 300 + time.Duration(rand.Intn(150)) * time.Microsecond
+
+	if rf.state != LEADER && time.Now().Sub(rf.lastHeartBeat) >= electionTimeout {
```

Comment:
[nit] Maybe we could try real time leader election right after timeout rather than periodically check
But we could try your way first, if there is a problem we could try real time way

Reply:
Done.